### PR TITLE
fn: more fn goodies

### PR DESCRIPTION
**Pull request by ProofOfKeags**
## Change Description
We add `fn` equivalents of the following Haskell functions:
- [`sequenceA`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:sequenceA) renamed to `Collect`, both the Option and Result variants
- [`traverse`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:traverse), both Option and Result variants
- [`mapMaybe`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Maybe.html#v:mapMaybe), renamed to FilterMap
- [`catMaybes`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Maybe.html#v:catMaybes), renamed to TrimNones
- [`note`](https://hackage.haskell.org/package/errors-2.3.0/docs/Control-Error-Util.html#v:note)
- [`hush`](https://hackage.haskell.org/package/errors-2.3.0/docs/Control-Error-Util.html#v:hush)

These are additions to the API that I discovered were missing when trying to do a refactor.

## Steps to Test
I can add tests if need be but for the most part I think the functions themselves are mostly checked by the type system. Thoughts appreciated.

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: fn/slice.go
Comment (carlaKC) [inline fn/slice.go:83]: nittynit here and above: `var somes []A` will save us an allocation in the case where every element is none.

File: fn/slice.go
Comment (carlaKC) [inline fn/slice.go:77]: surprised the linter isn't complaining about newline before return - did we turn that off 🤔

File: fn/slice.go
Comment (carlaKC) [inline fn/slice.go:302]: nit: unnecessary newline

File: fn/slice.go
Comment (carlaKC) [inline fn/slice.go:290]: For my own context, what's an example use case for this?

File: fn/option.go
Comment (carlaKC) [inline fn/option.go:242]: Just pass in `e error`?

File: fn/result.go
Comment (carlaKC) [inline fn/result.go:85]: Identical to the `Option` helper above?

File: fn/result.go
Comment (carlaKC) [inline fn/result.go:92]: Remove rather than deprecate with a comment?

File: fn/option.go
Comment (carlaKC) [inline fn/option.go:222]: nit: `OptionToLeft` and `OptionToRight` change could be in own commit.

File: fn/slice.go
Comment (carlaKC) [inline fn/slice.go:341]: Tests for the functions in this commit?

**Review by carlaKC — COMMENTED** [review]
Nice additions, no major comments!

File: fn/slice.go
Comment (ProofOfKeags) [reply fn/slice.go:83]: Rare but fair.

File: fn/option.go
Comment (ProofOfKeags) [reply fn/option.go:242]: The main use case here is when something that ordinarily returns None needs a new error attached to it. We *can* pass an error here but I think it's less useful than this. Perhaps I can do both `Note` and `Notef` variants.

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:85]: Yes this is an alias.

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:92]: I would need to do that in the v2 push and it's already on my list. Other packages/projects depend on this package and so breaking the API without warning isn't OK.

File: fn/slice.go
Comment (ProofOfKeags) [reply fn/slice.go:290]: Let's say you want to dispatch a number of operations that all return an option, but you only want to succeed and move the computation forward if they *all* succeed. This allows us do do a functor inversion between the list functor and the option functor in a way that ensures all of the options came back as some. This is distinct from something like TrimNones where we always succeed but prune the failed computations.

CollectResults will be more useful in practice but the semantics of Result and Option usually mirror each other. The only difference between them is that Option is a monad with a *canonical* annihilator and Result is not since there is no sensible "default error". All of their other semantics are identical though.

File: fn/option.go
Comment (carlaKC) [reply fn/option.go:242]: Ah gotcha, don't have very strong feelings here so happy to leave as-is 👍

Comment (ProofOfKeags) [reply fn/option.go]: I split it into Note and Notef

File: fn/option.go
Comment (carlaKC) [inline fn/option.go:68]: `require.False(o.isSome)`

File: fn/slice_test.go
Comment (carlaKC) [inline fn/slice_test.go:482]: debug print? here and in a few places below

File: fn/fn.go
Comment (carlaKC) [inline fn/fn.go:26]: Unnecessary change?

File: fn/result.go
Comment (carlaKC) [reply fn/result.go:92]: > I would need to do that in the v2 push

Meaning, a v2 of this package with breaking changes?

File: fn/result.go
Comment (carlaKC) [reply fn/result.go:85]: If it's an alias shouldn't it call `Option` rather than duplicating its logic? 

What's the motivation for an alias for this vs just renaming `Option`?

**Review by carlaKC — APPROVED** [review]
lgtm, just nits

File: fn/fn.go
Comment (ProofOfKeags) [reply fn/fn.go:26]: Not *necessary* but is a large QOL boost. The A parameter can be inferred by the supplied argument, but the B parameter cannot. Go does not require all type parameters to be specified if it can infer them from the arguments. However go *does NOT* possess the ability to do return type inference based off of the calling context. So by reversing the order of the type parameters here we now *only* need to specify B in most places since A will be inferred.

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:92]: Correct.

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:85]: Because I intend to remove Option in a later (breaking) PR.

File: fn/option.go
Comment (ProofOfKeags) [reply fn/option.go:68]: Won't that make the error message that appears in a test less helpful though?

Comment (carlaKC) [reply fn/option.go:68]: Require throws the line that failed so should be enough and it allows setting a message: 
`require.False(t, o.isSome, "Option %T is None()", o.some)`

File: fn/fn.go
Comment (carlaKC) [reply fn/fn.go:26]: TIL!

File: fn/result.go
Comment (carlaKC) [reply fn/result.go:85]: Ah ok, definitely missing some context on the longer term plan for a v2 version of this. 

Maybe add a note that `Option` is deprecated as done for `WhenResult` then?

**Comment by ProofOfKeags** [comment]
This PR has transformed into the breaking change PR we have been planning for some time. Micro label has been stripped. Merging this means we must push a tag for fn v2.0.0

File: fn/result.go
Comment (carlaKC) [inline fn/result.go:88]: `ToOption`? `Hush` doesn't intuitively tell me what we're doing here

File: fn/result_test.go
Comment (carlaKC) [inline fn/result_test.go:29]: 🥇

File: fn/either.go
Comment (carlaKC) [inline fn/either.go:67]: The old names `RightToOption`/`LeftToOption` are more intuitive imo.

File: fn/result.go
Comment (carlaKC) [inline fn/result.go:66]: Can be squashed with the commit where `Hush` is added.

File: fn/result.go
Comment (carlaKC) [inline fn/result.go:192]: Recognize that this is quite subjective nitting, but this is quite an odd function name imo - `ToResultFn`?

If there's some canonical terminology for this that I'm missing I'll hush, but at least from a rustlang perspective I don't think there is?

**Review by carlaKC — APPROVED** [review]
Some name nitpicking, because it wouldn't be breaking refactor without some totally subjective opinions. Overall code lgtm!

Approving with the caveat that I don't have the full context on how this package is being used (both inside of LND and outside of it). 

I lean towards the approach of adding functionality as we need it (rather than wholesale PRs to extend the `fn` package), but acknowledge that this is a flawed view because of the chicken-and-egg issue that:
* Package is less likely to be used if it doesn't have useful fundamentals
* If it's not being used, people won't add said useful fundamentals

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:88]: If a second opinion agrees, I'll change it.

File: fn/either.go
Comment (ProofOfKeags) [reply fn/either.go:67]: If a second opinion agrees, I'll change it.

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:192]: [liftA2](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:liftA2) is a general combinator for lifting arbitrary binary operations over any applicative functor context. The applicative here is `Result` and the intuition is that you are "lifting" some binary ("2") function `f` over the Applicative ("A"). Since go does not support HKTs, it is impossible to define this combinator once for all Applicatives, so I postfix the name "LiftA2" with the functor it is monomorphized to (in this case "Result").

There are corresponding functions for all Applicative functors, notably: Option, Result, List (two kinds: cartesian product and zip lists), Futures, functions *from* some fixed type, and any composition of these functors, since Applicatives are closed under composition (if `F` and `G` are both applicative functors then `F . G` and `G . F` are themselves applicative functors).

As with most things I'm not attached to the names, but I did choose this name for a reason. You are not the first person to provide this feedback and so I'm happy to rename this lifting combinator, but I think it's important that the name chosen is consistent for all of its instantiations (There is currently a `LiftA2Option` that would also need to be renamed).

**Comment by ProofOfKeags** [comment]
> I lean towards the approach of adding functionality as we need it (rather than wholesale PRs to extend the fn package)

The tools I include have all been directly motivated by code I witnessed in LND that would benefit from their inclusion, AND they all have direct corollaries in haskell's *standard* library. These aren't just toys I thought up from nowhere.

Adding functionality "as we need" is, in practice, just "continue to do the Sysiphean boulder push" since it is rare for people to notice and re-derive functional patterns, however, when they already exist, people will use them without issue.

The curse of truly general code is that it's generality makes it simultaneously more useful AND *less obviously* useful.

File: fn/result.go
Comment (carlaKC) [reply fn/result.go:192]: > As with most things I'm not attached to the names, but I did choose this name for a reason. You are not the first person to provide this feedback and so I'm happy to rename this lifting combinator, but I think it's important that the name chosen is consistent for all of its instantiations (There is currently a LiftA2Option that would also need to be renamed).

Happy to leave name as-is and keep it consistent, thanks for explaining the reasoning!

**Comment by carlaKC** [comment]
> since it is rare for people to notice and re-derive functional patterns, however, when they already exist, people will use them without issue.

Totally valid point / approach 👍 

My comment was just to note that I defer to the best judgement of the author/second reviewer here because I haven't done the work of taking a look at LND / common patterns that could use these tools, not that they aren't going to be useful!

File: fn/slice.go
Comment (starius) [inline fn/slice.go:20]: IIUC this change is needed, because the predicate is often a closure and it is more convenient to pass it as last argument. Please add the reason to have the change to the commit message.

File: fn/either.go
Comment (starius) [reply fn/either.go:67]: +1 for old names `RightToOption` and `LeftToOption`

File: fn/result.go
Comment (starius) [reply fn/result.go:192]: I like LiftA2Result name

**Review by starius — APPROVED** [review]
LGTM 🏆
Great work!
I left several comments and proposals.

File: fn/slice.go
Comment (ProofOfKeags) [reply fn/slice.go:71]: This may lead to gross over allocation which I think is the wrong choice. If the predicate passes very rarely then the size of the return value will be far under capacity. CPU is cheaper than memory

File: fn/slice.go
Comment (ProofOfKeags) [reply fn/slice.go:85]: See above

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:81]: If that's the case I'd recommend ditching the methods since they are less useful (they are constrained to endomorphic functions)

File: fn/slice.go
Comment (ProofOfKeags) [reply fn/slice.go:20]: Yes we very often define the functional arguments on the fly making them read better as the last argument.

File: fn/option.go
Comment (ProofOfKeags) [reply fn/option.go:255]: Fair enough. This is a very FP explanation. I'll make it more accessible.

File: fn/result.go
Comment (ziggie1984) [reply fn/result.go:81]: I would keep it

**Review by ziggie1984 — APPROVED** [review]
LGTM

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:129]: I'll add a test. Once again I'm still skeptical of making the testing object into an interface.

File: fn/slice.go
Comment (ProofOfKeags) [reply fn/slice.go:330]: Looks like the linter doesn't care if the function is sufficiently small but I fixed this anyway.

File: fn/result.go
Comment (ProofOfKeags) [reply fn/result.go:66]: I think I'd rather keep the commits separate since this one is breaking.

**Review by starius — APPROVED** [review]
LGTM

**Pull request merged**
